File: baekjoon/2961.cpp
#include <iostream>
#include <vector>
#include <math.h>
using namespace std;

// 도영이가 만든 맛있는 음식

struct flavor {
	int s;
	int b;
	bool isUsed;
};

int N;
vector <flavor> fl;
int recipe = 999999;

void findRecipe(int count, int pos, int s, int b)
{
	if (count > 0) {
		int diff = abs(s - b); // 신맛과 쓴맛의 차
		recipe = recipe > diff ? diff : recipe; // 기존 레시피보다 diff가 작으면 diff로 레시피 변경
	}

	for (int i = pos; i < N; i++)
	{
		if (!fl[i].isUsed)
		{
			fl[i].isUsed = true;
			findRecipe(count + 1, i, s * fl[i].s, b + fl[i].b); // 이 재료를 썼을 때의 최소값
			fl[i].isUsed = false;
		}
	}
}

int main(void)
{
	cin >> N;

	for (int i = 0; i < N; i++)
	{
		int s, b;
		cin >> s >> b;
		fl.push_back({ s, b, false });
	}

	findRecipe(0, 0, 1, 0);

	cout << recipe;

	return 0;
}

File: baekjoon/2961.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.lang.Integer.min
import java.util.*
import kotlin.collections.ArrayList
import kotlin.math.abs

// 도영이가 만든 맛있는 음식

var N: Int = 0
var answer: Int = Int.MAX_VALUE
val dishes = ArrayList<Dish>()
lateinit var isUsed: BooleanArray

fun checkDish(count: Int, index: Int, sour: Int, bitter: Int) {
    if(count > 0) {
        answer = min(answer, abs(sour - bitter))
    }

    if(count < dishes.size) {
        for (i in index..dishes.lastIndex) {
            if(!isUsed[i]) {
                isUsed[i] = true
                checkDish(count + 1, i,sour * dishes[i].S, bitter + dishes[i].B)
                isUsed[i] = false
            }
        }
    }
}

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    var token = StringTokenizer(br.readLine())
    N = token.nextToken().toInt()
    isUsed = BooleanArray(N) {false}

    for (i in 1..N) {
        token = StringTokenizer(br.readLine())
        dishes.add(Dish(S = token.nextToken().toInt(), B = token.nextToken().toInt()))
    }

    checkDish(0, 0,1, 0)
    print(answer)
}

data class Dish (
        val S: Int = 0, // 신맛
        val B: Int = 0 // 쓴맛
)